lib/puppet/provider/azure_vm/azure_arm.rb

```diff
@@ -83,42 +83,50 @@ def default_based_on_resource_group(value)
 
   def create # rubocop:disable Metrics/AbcSize
     Puppet.info("Creating #{resource[:name]}")
-    create_vm({
-      # required
-      name: resource[:name],
-      image: resource[:image],
-      location: resource[:location],
-      size: resource[:size],
-      user: resource[:user],
-      password: resource[:password],
-      resource_group: resource[:resource_group],
-      # type defaults
-      storage_account_type: resource[:storage_account_type],
-      os_disk_caching: resource[:os_disk_caching],
-      os_disk_create_option: resource[:os_disk_create_option],
-      os_disk_vhd_container_name: resource[:os_disk_vhd_container_name],
-      dns_servers: resource[:dns_servers],
-      public_ip_allocation_method: resource[:public_ip_allocation_method],
-      virtual_network_address_space: resource[:virtual_network_address_space],
-      subnet_name: resource[:subnet_name],
-      subnet_address_prefix: resource[:subnet_address_prefix],
-      private_ip_allocation_method: resource[:private_ip_allocation_method],
-      # provider defaults recreate the defaults from the Azure Portal
-      storage_account: default_based_on_resource_group(resource[:storage_account]),
-      os_disk_name: default_to_name(resource[:os_disk_name]),
-      os_disk_vhd_name: default_to_name(resource[:os_disk_vhd_name]),
-      dns_domain_name: default_to_simple_name(resource[:dns_domain_name]),
-      public_ip_address_name: default_to_name(resource[:public_ip_address_name]),
-      virtual_network_name: default_to_resource_group(resource[:virtual_network_name]),
-      ip_configuration_name: default_to_name(resource[:ip_configuration_name]),
-      network_interface_name: default_based_on_name(resource[:network_interface_name]),
-    })
+    begin
```

Comment:
All the error handling should happen as near as possible to the API call. If you look at https://github.com/puppetlabs/puppetlabs-azure/blob/master/lib/puppet_x/puppetlabs/azure/provider_arm.rb , there are a log of `promise.value!` calls that effect an API call that could potentially throw.